passthru FB_face_tracking2

galister opened the pull request:
FB_face_tracking2 support got merged into Monado just now.

We only implement the VISUAL tracking method, because that's the one that makes sense.

Tested with OscAvMgr + VRChat.

xytovl commented on common/wivrn_packets.h:
```diff
@@ -185,6 +186,14 @@ struct hand_tracking
 
+struct fb_face2
+{
+	std::array<float, 70> weights;
```
We could use `XR_FACE_EXPRESSION2_COUNT_FB` if we make sure we have a recent enough OpenXR

xytovl commented on client/scenes/stream_tracking.cpp:
```diff
@@ -243,6 +244,12 @@ void scenes::stream::tracking()
 
+					if (application::get_fb_face_tracking2_supported())
+					{
+						application::get_fb_face_tracker2().get_weights(t0 + Δt, &fb_face2);
```
Either we do the extrapolation + latest (Δt = 0 and Δt = prediction) and server side needs to use a history system to store/retrieve the values, or we only query one time.

What I understand from current code is that it will get 2 samples per loop, which might get reordered, then server only stores one.

galister replied:
We only get 1 sample and store that. But maybe extrapolating on the server side would make sense.

xytovl replied:
`get_weights` is inside the `for (XrDuration Δt = 0; Δt <= prediction; Δt += std::max<XrDuration>(1, prediction))` loop, it will be executed twice

galister replied:
so i guess the battery check also needs to be moved out?

xytovl commented on client/scenes/stream_tracking.cpp:
```diff
@@ -243,6 +244,12 @@ void scenes::stream::tracking()
+					if (application::get_fb_face_tracking2_supported())
+					{
+						application::get_fb_face_tracker2().get_weights(t0 + Δt, &fb_face2);
+						network_session->send_stream(fb_face2);
```
Timer should be paused when sending, I guess manually doing this everytime doesn't scale, we can make a macro so that future additions won't miss it.

Rationale: socket sometimes block (queue full? radio error?), and mess up time measurement. On some runtimes, locating spaces is heavy on CPU time and it is not possible to run the loop at 1kHz rate. Lowering the sampling rate helps with thermals and is a bit better. We don't want to lower the rate for occasional network spikes.

xytovl commented on server/driver/wivrn_hmd.cpp:
```diff
@@ -274,6 +292,16 @@ void wivrn_hmd::update_battery(const from_headset::battery & new_battery)
 	battery = new_battery;
 }
 
+void wivrn_hmd::update_fb_face2(const from_headset::fb_face2 & new_face)
+{
+	memcpy(&face.face_expression_set2_fb.weights, new_face.weights.data(), sizeof(float) * new_face.weights.size());
+	memcpy(&face.face_expression_set2_fb.confidences, new_face.confidences.data(), sizeof(float) * new_face.confidences.size());
+	face.face_expression_set2_fb.is_valid = new_face.is_valid;
+	face.face_expression_set2_fb.is_eye_following_blendshapes_valid = new_face.is_eye_following_blendshapes_valid;
+	face.face_expression_set2_fb.data_source = XRT_FACE_TRACKING_DATA_SOURCE2_VISUAL_FB;
+	face.face_expression_set2_fb.sample_time_ns = os_monotonic_get_ns();
+}
+
```
This is not thread safe, and considering the comment on the prediction time, I think deriving from `history` could be better.

galister commented:
I think the Wireshark pipeline's missing the openxr header.

Either way, I'll look at using history, but I've a busy week so it might take some time.

galister commented:
nvm, libopenxr-dev in jammy is just too old

galister commented:
Since last review:
- I moved the battery check outside the predict loop
- I rewrote the face tracker to use history
- Face tracker is now its own device
- It works and it's happy

However, I will need help with:
- Getting an up-to-date `openxr.h` into the Wireshark dissector
- Figure out why the XRT defines don't work in the Flatpak action

xytovl commented on server/driver/weights_list.h:
```diff
@@ -0,0 +1,59 @@
+		return result;
+	}
+
+	static std::array<float, NumWeights> extrapolate(const std::array<float, NumWeights> & a, const std::array<float, NumWeights> & b, uint64_t ta, uint64_t tb, uint64_t t)
+	{
+		float h = (tb - ta) / 1.e9;
+
+		std::array<float, NumWeights> result = t < ta ? a : b;
+		for (size_t i = 0; i < NumWeights; i++)
+		{
+			auto vel = (b[i] - a[i]) / h;
+			float dt = (t - tb) / 1.e9;
+
+			result[i] = std::clamp(result[i] + vel * dt, 0.0f, 1.0f);
+		}
+		return result;
+	}
```
It should be possible to not implement this by replacing  the `if (extrapolate)` with `if constexpr (extrapolate)`.

I don't know how much sense it makes to extrapolate expressions.

galister replied:
it's not being used right now as the underlying history has extrapolate off, but I implemented it in case it would be used for something else that needs extrapolation

xytovl commented on server/driver/wivrn_fb_face2_tracker.cpp:
```diff
@@ -0,0 +1,150 @@
+
+		XrTime predicted_time = ((XrTime)now) + last_prediction_offset;
+
+		out_value->face_expression_set2_fb.sample_time_ns = predicted_time;
```
The sample_time could be stored as data with the weights list, where interpolate/extrapolate would likely take the max value. No need to store it out of the history object.

I don't if there is a reason to split weights and confidences

galister replied:
the idea was to keep weights_list generic enough so that it could be used for a different face tracking extension down the line.

xytovl replied:
If you think it makes sense to have 2 objects, it's not a big problem, but the sample time should be part of the data

xytovl commented:
> However, I will need help with:
> 
> * Getting an up-to-date `openxr.h` into the Wireshark dissector check

I have pushed a commit with the cmake rules for dissector
